lib/fputs.js

```diff
@@ -0,0 +1,172 @@
+/**
+ * Fputs -- fast buffered line-at-a-time output
+ *
+ * Exports fputs() and write(), both buffer data and flush in the background.
+ * Fputs will buffer an unbounded amount of data; use drain() to throttle.
+ *
+ * Copyright (C) 2014-2015 Andras Radics
+ * Licensed under the Apache License, Version 2.0
+ *
+ * 2014-09-14 - AR.
+ */
+
+var setImmediate = global.setImmediate || process.nextTick;
+
+module.exports = (function() {
+    'use strict';
+
+    /**
+     * Constructor, takes a writable (else a filename) and options.
+     *
+     * options:
+     *     writemode:   file open mode, default 'a'
+     *     writesize:   number of chars to write per chunk, default 200k
+     */
+    function Fputs( writable, opts ) {
+        if (!(this instanceof Fputs)) return new Fputs(writable, opts);
+
+        opts = opts || {};
+        if (typeof writable === 'string') {
+            // convert a string filename into a FileWriter writable
+            writable = new Fputs.FileWriter(writable, opts.writemode || 'a');
+        }
+        if (!writable) writable = process.stdout;
+
+        this.writable = writable;
+        this.writesize = opts.writesize || 204800;
+        this.datachunks = [];
+        this.unwrittenLength = 0;
+        this.writtenLength = 0;
+        this.resetCount = 0;
+        this._syncing = false;
+
+        this._error = null;
+        this.reportError = function(err) { if (!this._error) this._error = err; }
+        this.returnError = function() { var err = this._error; this._error = null; return err; }
+    }
+
+    /**
+     * Append a newline terminated string to the fifo.
+     */
+    Fputs.prototype.fputs = function fputs( str ) {
+        if (typeof str !== 'string') str = "" + str;
+        this.write((str[str.length - 1] === "\n") ? str : str + "\n");
+    }
+
+    /**
+     * Write bulk data to the target.  Newline termination is not checked.
+     */
+    Fputs.prototype.write = function write( str, callback ) {
+        if (typeof str !== 'string') str = "" + str;
+
+        // merge writes into this.writesize sized data chunks
+        // it is assumed that all writes end with a newline (not checked)
```

Comment:
they get concatenated and written in larger batches, as expected.  This is in fact how kds uses qfputs
(the newline separators between batches were added independently, for legibility)

The reason for the comment is that qfputs was designed for high-speed streaming of newline-terminated
strings, with each line written to the file atomically (ie, each record lands in the file with its newline terminator).
This invariant for the data consumer is maintained by the `fputs()` method, but `write()` ignores newlines
and thus could send partial records to the consumer of the data stream.

The note is a reminder for programmers who manually concatenate records that they must
ensure that records are batched in their entirety with their terminating newline for data consistency.